typesetters/base.lua

```diff
@@ -315,18 +322,147 @@ function typesetter:breakIntoLines (nodelist, breakWidth)
+    xOffset = 0
+  else
+    -- Computing italic correction is at best heuristics.
+    -- The strong assumption is that italic is slanted to the right.
+    -- Thus, the part of the character that goes beyond its width is usually
+    -- maximal at the top of the glyph.
+    -- E.g. consider a "f", that would be the top hook extent.
+    -- Pathological cases exist, such as fonts with a Q with a long tail,
+    -- but these will rarely occur in usual languages. For instance, Klingon's
+    -- "QaQ" might be an issue, but there's not much we can do...
+    -- Another assumption is that we can distribute that extent in proportion
+    -- with the next character's height.
+    -- This might not work that well with non-Latin scripts.
+    local d = precShape.glyphWidth + precShape.x_bearing
+    local delta = d > precShape.width and d - precShape.width or 0
+    xOffset = precShape.height <= curShape.height
+      and delta
+      or delta * curShape.height / precShape.height
+  end
+  return xOffset
+end
+
+local function toItalicCorrection (precShape, curShape)
+  if not SILE.settings:get("typesetter.italicCorrection") then return end
+  local xOffset
+  if not curShape or not precShape then
+    xOffset = 0
+  else
+    -- Same assumptions as fromItalicCorrection(), but on the starting side of
+    -- the glyph.
+    local d = curShape.x_bearing
+    local delta = d < 0 and -d or 0
+    xOffset = precShape.depth >= curShape.depth
+      and delta
+      or delta * precShape.depth / curShape.depth
+  end
+  return xOffset
+end
+
+local function isItalicLike(nnode)
+  -- We could do...
+  --  return nnode and string.lower(nnode.options.style) == "italic"
+  -- But it's probably more robust to use the italic angle, so that
+  -- thin italic, oblique or slanted fonts etc. may work too.
+  local ot = require("core.opentype-parser")
+  local face = SILE.font.cache(nnode.options, SILE.shaper.getFace)
+  local font = ot.parseFont(face)
+  return font.post.italicAngle ~= 0
+end
+
+function typesetter.shapeAllNodes (_, nodelist, inplace)
+  inplace = SU.boolean(inplace, true) -- Compatibility with earlier versions
+  local newNodelist = {}
+  local prec
+  local precShapedNodes
+  for _, current in ipairs(nodelist) do
+    if current.is_unshaped then
+      local shapedNodes = current:shape()
+
+      if SILE.settings:get("typesetter.italicCorrection") and prec then
+        local itCorrOffset
+        if isItalicLike(prec) and not isItalicLike(current) then
+          local precShape = getLastShape(precShapedNodes)
+          local curShape = getFirstShape(shapedNodes)
+          itCorrOffset = fromItalicCorrection(precShape, curShape)
+        elseif not isItalicLike(prec) and isItalicLike(current) then
+          local precShape = getLastShape(precShapedNodes)
+          local curShape = getFirstShape(shapedNodes)
+          itCorrOffset = toItalicCorrection(precShape, curShape)
+        end
+        if itCorrOffset then
```

Comment:
Maybe `and itCorrOffset ~= 0`  --- Kind of useless to insert a zero-width kern here.

Reply:
Useless indeed ... as long as we're not also breaking line breaking. Are we supposed to be inhibiting breaks at these points, and if so are we still?

Reply:
In the case of the french punctuation spaces, they are non-breakable, so it's fine whatever we do.

But in the general case, your are guessing right... There's a problem indeed :(

![image](https://user-images.githubusercontent.com/18075640/235358325-08ce8ee2-2529-45b0-9389-e1c548d22d2b.png)

(I tweaked the correction kern output routine so that is shows a red box)

Ah. A break does not occur any longer at the regular space glue. (If it did, we could still have a small justification issue due to the correction kern in some extreme cases, but that could be _quite_ acceptable... But see below).

It's fixable:

![image](https://user-images.githubusercontent.com/18075640/235358801-9efec467-715f-41a6-b32a-b4bd279d95a2.png)

(Now, red = kern correction, blue = glue correction)

The implication is that depending on the presence of a (space) glue, the italic correction should be a glue rather than a space. (And also, end-of-line glues are kinda skipped in the line ratio, so it wouldn't affect justification it seems).

Reply:
Fixed.